File: 백준/Gold/1520. 내리막 길/내리막 길.py
# https://www.acmicpc.net/problem/1520

N, M = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]
dp = [[-1 for _ in range(M)] for _ in range(N)]

dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]

def in_range(r, c):
    return 0 <= r < N and 0 <= c < M
# 항상 높이가 더 낮은 지점으로만 이동하여 목표지점으로 이동!
# dpf(r, c): (0, 0)에서 (r, c)까지의 경우의 수 > DP를 생각할 수 있다!!
# dpf(r, c) = dpf(pre_r, pre_c) + 1 where board[pre_r][pre_c] > board[r][c]
def dpf(r: int, c: int):
    if dp[r][c] != -1:
        return dp[r][c]

    ret = 0
    for k in range(4):
        pre_r, pre_c = r + dr[k], c + dc[k]
        if in_range(pre_r, pre_c) and board[pre_r][pre_c] > board[r][c]:
            ret += dpf(pre_r, pre_c)
            # print(f"pre_r: {pre_r}, pre_c: {pre_c}, board: {board[pre_r][pre_c]}")

    # 시작점은 (0, 0)
    if r == 0 and c == 0:
        return 1

    dp[r][c] = ret

    return ret

print(dpf(N - 1, M - 1)) 

File: 백준/Gold/1520. 내리막 길/내리막 길.java
import java.util.*;

class Main
{	private static int n, m;
	private static int[][] board, dp;
	
	private static int[] dr = new int[]{-1, 1, 0, 0};
	private static int[] dc = new int[]{0, 0, -1, 1};
	
	public static void main (String[] args)
	{
		Scanner sc = new Scanner(System.in);
		n = sc.nextInt();
		m = sc.nextInt();
		
		board = new int[n][m];
		dp = new int[n][m];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				int x = sc.nextInt();
				board[i][j] = x;
				dp[i][j] = -1;
			}
		}
		
		System.out.print(dpf(n - 1, m - 1));
		
	}
	
	private static int dpf(int r, int c) {
		if (dp[r][c] != -1) {
			return dp[r][c];
		}
		
		int ret = 0;
		for (int d = 0; d < 4; d++) {
			int preR = r + dr[d];
			int preC = c + dc[d];
			if (inRange(preR, preC) && board[preR][preC] > board[r][c]) {
				ret += dpf(preR, preC);
			}
		}
		
		if (r == 0 && c == 0) {
			return 1;
		}
		
		dp[r][c] = ret;
		
		return ret;
	}
	
	private static boolean inRange(int r, int c) {
		return 0 <= r && r < n && 0 <= c && c < m;
	}
}